// @helius-labs/helius-sdk/dist/esm/wallet/client.eager.js
import { getIdentity } from './getIdentity.js';
import { getBatchIdentity } from './getBatchIdentity.js';
import { getBalances } from './getBalances.js';
import { getHistory } from './getHistory.js';
import { getTransfers } from './getTransfers.js';
import { getFundedBy } from './getFundedBy.js';

/**
 * Create a Wallet API client with eager loading
 *
 * All methods are imported immediately. This client is primarily used for testing
 * where lazy loading is not needed.
 *
 * @param apiKey - Helius API key
 * @returns WalletClient instance with eagerly-loaded methods
 *
 * @example
 * ```ts
 * import { makeWalletClientEager } from "helius-sdk/wallet/client.eager";
 *
 * const walletClient = makeWalletClientEager("your-api-key");
 * const identity = await walletClient.getIdentity({ wallet: "..." });
 * ```
 */
const makeWalletClientEager = (apiKey, userAgent) => ({
    getIdentity: (p) => getIdentity(apiKey, p, userAgent),
    getBatchIdentity: (p) => getBatchIdentity(apiKey, p, userAgent),
    getBalances: (p) => getBalances(apiKey, p, userAgent),
    getHistory: (p) => getHistory(apiKey, p, userAgent),
    getTransfers: (p) => getTransfers(apiKey, p, userAgent),
    getFundedBy: (p) => getFundedBy(apiKey, p, userAgent),
});

export { makeWalletClientEager };
