// @helius-labs/helius-sdk/dist/esm/wallet/client.js
/**
 * Create a Wallet API client with lazy loading
 *
 * Methods are dynamically imported on first use to optimize bundle size.
 * This is the default client used by `createHelius()`.
 *
 * @param apiKey - Helius API key
 * @returns WalletClient instance with lazy-loaded methods
 *
 * @example
 * ```ts
 * import { makeWalletClient } from "helius-sdk/wallet/client";
 *
 * const walletClient = makeWalletClient("your-api-key");
 * const identity = await walletClient.getIdentity({ wallet: "..." });
 * ```
 */
const makeWalletClient = (apiKey, userAgent) => ({
    getIdentity: async (p) => (await import('./getIdentity.js')).getIdentity(apiKey, p, userAgent),
    getBatchIdentity: async (p) => (await import('./getBatchIdentity.js')).getBatchIdentity(apiKey, p, userAgent),
    getBalances: async (p) => (await import('./getBalances.js')).getBalances(apiKey, p, userAgent),
    getHistory: async (p) => (await import('./getHistory.js')).getHistory(apiKey, p, userAgent),
    getTransfers: async (p) => (await import('./getTransfers.js')).getTransfers(apiKey, p, userAgent),
    getFundedBy: async (p) => (await import('./getFundedBy.js')).getFundedBy(apiKey, p, userAgent),
});

export { makeWalletClient };
