// @helius-labs/helius-sdk/dist/cjs/wallet/client.eager.cjs
'use strict';

var getIdentity = require('./getIdentity.cjs');
var getBatchIdentity = require('./getBatchIdentity.cjs');
var getBalances = require('./getBalances.cjs');
var getHistory = require('./getHistory.cjs');
var getTransfers = require('./getTransfers.cjs');
var getFundedBy = require('./getFundedBy.cjs');

/**
 * Create a Wallet API client with eager loading
 *
 * All methods are imported immediately. This client is primarily used for testing
 * where lazy loading is not needed.
 *
 * @param apiKey - Helius API key
 * @returns WalletClient instance with eagerly-loaded methods
 *
 * @example
 * ```ts
 * import { makeWalletClientEager } from "helius-sdk/wallet/client.eager";
 *
 * const walletClient = makeWalletClientEager("your-api-key");
 * const identity = await walletClient.getIdentity({ wallet: "..." });
 * ```
 */
const makeWalletClientEager = (apiKey, userAgent) => ({
    getIdentity: (p) => getIdentity.getIdentity(apiKey, p, userAgent),
    getBatchIdentity: (p) => getBatchIdentity.getBatchIdentity(apiKey, p, userAgent),
    getBalances: (p) => getBalances.getBalances(apiKey, p, userAgent),
    getHistory: (p) => getHistory.getHistory(apiKey, p, userAgent),
    getTransfers: (p) => getTransfers.getTransfers(apiKey, p, userAgent),
    getFundedBy: (p) => getFundedBy.getFundedBy(apiKey, p, userAgent),
});

exports.makeWalletClientEager = makeWalletClientEager;
