// @helius-labs/helius-sdk/dist/cjs/wallet/client.cjs
'use strict';

/**
 * Create a Wallet API client with lazy loading
 *
 * Methods are dynamically imported on first use to optimize bundle size.
 * This is the default client used by `createHelius()`.
 *
 * @param apiKey - Helius API key
 * @returns WalletClient instance with lazy-loaded methods
 *
 * @example
 * ```ts
 * import { makeWalletClient } from "helius-sdk/wallet/client";
 *
 * const walletClient = makeWalletClient("your-api-key");
 * const identity = await walletClient.getIdentity({ wallet: "..." });
 * ```
 */
const makeWalletClient = (apiKey, userAgent) => ({
    getIdentity: async (p) => (await Promise.resolve().then(function () { return require('./getIdentity.cjs'); })).getIdentity(apiKey, p, userAgent),
    getBatchIdentity: async (p) => (await Promise.resolve().then(function () { return require('./getBatchIdentity.cjs'); })).getBatchIdentity(apiKey, p, userAgent),
    getBalances: async (p) => (await Promise.resolve().then(function () { return require('./getBalances.cjs'); })).getBalances(apiKey, p, userAgent),
    getHistory: async (p) => (await Promise.resolve().then(function () { return require('./getHistory.cjs'); })).getHistory(apiKey, p, userAgent),
    getTransfers: async (p) => (await Promise.resolve().then(function () { return require('./getTransfers.cjs'); })).getTransfers(apiKey, p, userAgent),
    getFundedBy: async (p) => (await Promise.resolve().then(function () { return require('./getFundedBy.cjs'); })).getFundedBy(apiKey, p, userAgent),
});

exports.makeWalletClient = makeWalletClient;
