// @helius-labs/helius-sdk/dist/cjs/websockets/wsAsync.cjs
'use strict';

const importWs = async () => (await import('@solana/kit')).createSolanaRpcSubscriptions;
/** Create a promisified WebSocket RPC subscription client. */
const makeWsAsync = (wsUrl) => {
    let _raw;
    const raw = async () => {
        if (_raw)
            return _raw;
        const ctor = await importWs();
        _raw = ctor(wsUrl);
        return _raw;
    };
    return {
        logsNotifications: (filter, config) => raw().then((r) => r.logsNotifications(filter, config)),
        slotNotifications: (config) => raw().then((r) => r.slotNotifications(config)),
        signatureNotifications: (sig, config) => raw().then((r) => r.signatureNotifications(sig, config)),
        programNotifications: (...args) => raw().then((r) => r.programNotifications(...args)),
        accountNotifications: (...args) => raw().then((r) => r.accountNotifications(...args)),
        close() {
            // `@solana/kit` exposes `.dispose()`; but we fall back to `.close()` or noop
            if (_raw && typeof _raw.dispose === "function") {
                _raw.dispose();
            }
            else if (_raw && typeof _raw.close === "function") {
                _raw.close();
            }
            _raw = undefined;
        },
    };
};

exports.makeWsAsync = makeWsAsync;
