// @helius-labs/helius-sdk/dist/cjs/transactions/client.cjs
'use strict';

var getComputeUnits = require('./getComputeUnits.cjs');
var createSmartTransaction = require('./createSmartTransaction.cjs');
var sendSmartTransaction = require('./sendSmartTransaction.cjs');
var pollTransactionConfirmation = require('./pollTransactionConfirmation.cjs');
var broadcastTransaction = require('./broadcastTransaction.cjs');
var createSmartTransactionWithTip = require('./createSmartTransactionWithTip.cjs');
var sendTransactionWithSender = require('./sendTransactionWithSender.cjs');
var sendTransaction = require('./sendTransaction.cjs');

const makeTxHelpersLazy = (rpc, getPriorityFeeEstimate, rpcSubscriptions) => {
    const getComputeUnits$1 = getComputeUnits.makeGetComputeUnits(rpc);
    const { create } = createSmartTransaction.makeCreateSmartTransaction({
        raw: rpc,
        getComputeUnits: getComputeUnits$1,
        getPriorityFeeEstimate,
    });
    const { create: createSmartTransactionWithTip$1 } = createSmartTransactionWithTip.makeCreateSmartTransactionWithTip(create);
    const pollTransactionConfirmation$1 = pollTransactionConfirmation.makePollTransactionConfirmation(rpc);
    const broadcastTransaction$1 = broadcastTransaction.broadcastTransactionFactory(rpc);
    const { send } = sendSmartTransaction.makeSendSmartTransaction({
        raw: rpc,
        rpcSubscriptions,
        createSmartTransaction: create,
    });
    const { send: sendWithSender } = sendTransactionWithSender.makeSendTransactionWithSender({
        raw: rpc,
        createSmartTransactionWithTip: createSmartTransactionWithTip$1,
    });
    const { send: sendTransaction$1 } = sendTransaction.makeSendTransaction(rpc);
    return {
        getComputeUnits: getComputeUnits$1,
        pollTransactionConfirmation: pollTransactionConfirmation$1,
        broadcastTransaction: broadcastTransaction$1,
        createSmartTransaction: create,
        sendSmartTransaction: send,
        sendTransactionWithSender: sendWithSender,
        sendTransaction: sendTransaction$1,
    };
};

exports.makeTxHelpersLazy = makeTxHelpersLazy;
