// @helius-labs/helius-sdk/dist/cjs/transactions/sendViaSender.cjs
'use strict';

var kit = require('@solana/kit');
var types = require('./types.cjs');

/**
 * POST a base‑64 transaction to the chosen Sender region.
 * Sender mandates skipPreflight = true and maxRetries = 0
 *
 * Returns the transaction signature
 */
const sendViaSender = async (tx64, region = "Default", swqosOnly = false) => {
    const endpoint = swqosOnly
        ? `${types.senderFastUrl(region)}?swqos_only=true`
        : types.senderFastUrl(region);
    const res = await fetch(endpoint, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
            jsonrpc: "2.0",
            id: Date.now().toString(),
            method: "sendTransaction",
            params: [
                tx64,
                { encoding: "base64", skipPreflight: true, maxRetries: 0 },
            ],
        }),
    });
    /** Handle HTTP‑level failures early */
    if (!res.ok) {
        const text = await res.text();
        throw new Error(`Sender HTTP ${res.status}: ${text.slice(0, 200)}`);
    }
    const body = await res.json();
    if (typeof body === "string")
        return kit.signature(body);
    if (body && typeof body === "object") {
        if ("error" in body && body.error)
            throw new Error(JSON.stringify(body.error));
        if ("result" in body && typeof body.result === "string")
            return kit.signature(body.result);
    }
    throw new Error(`Unexpected Sender response: ${JSON.stringify(body).slice(0, 200)}`);
};

exports.sendViaSender = sendViaSender;
