// @helius-labs/helius-sdk/dist/cjs/transactions/types.cjs
'use strict';

/**
 * Regional Helius sender endpoints for SWQOS transaction submission.
 *
 * `Default` uses HTTPS and is suitable for frontend/browser applications.
 * All other regions use HTTP and are intended for server-side use.
 */
const SENDER_ENDPOINTS = {
    /** Global HTTPS endpoint — suitable for frontend applications. */
    Default: "https://sender.helius-rpc.com",
    /** Salt Lake City, USA. */
    US_SLC: "http://slc-sender.helius-rpc.com",
    /** Newark, USA. */
    US_EAST: "http://ewr-sender.helius-rpc.com",
    /** London, UK. */
    EU_WEST: "http://lon-sender.helius-rpc.com",
    /** Frankfurt, Germany. */
    EU_CENTRAL: "http://fra-sender.helius-rpc.com",
    /** Amsterdam, Netherlands. */
    EU_NORTH: "http://ams-sender.helius-rpc.com",
    /** Singapore. */
    AP_SINGAPORE: "http://sg-sender.helius-rpc.com",
    /** Tokyo, Japan. */
    AP_TOKYO: "http://tyo-sender.helius-rpc.com",
};
/** Build the `/fast` endpoint URL for a sender region. */
const senderFastUrl = (region) => `${SENDER_ENDPOINTS[region]}/fast`;
/** Build the `/ping` endpoint URL for a sender region. GET this URL to keep the connection warm. */
const senderPingUrl = (region) => `${SENDER_ENDPOINTS[region]}/ping`;
/** Tip account addresses used by the Helius sender infrastructure. */
const SENDER_TIP_ACCOUNTS = [
    "4ACfpUFoaSD9bfPdeu6DBt89gB6ENTeHBXCAi87NhDEE",
    "D2L6yPZ2FmmmTKPgzaMKdhu6EWZcTpLy1Vhx8uvZe7NZ",
    "9bnz4RShgq1hAnLnZbP8kbgBg1kEmcJBYQq3gQbmnSta",
    "5VY91ws6B2hMmBFRsXkoAAdsPHBJwRfBht4DXox3xkwn",
    "2nyhqdwKcJZR2vcqCyrYsaPVdAnFoJjiksCXJ7hfEYgD",
    "2q5pghRs6arqVjRvT5gfgWfWcHWmw1ZuCzphgd5KfWGJ",
    "wyvPkWjVZz1M8fHQnMMCDTQDbkManefNNhweYk5WkcF",
    "3KCKozbAaF75qEU33jtzozcJ29yJuaLJTy2jFdzUY8bT",
    "4vieeGHPYPG2MmyPRcYjdiDmmhN3ww7hsFNap8pVN3Ey",
    "4TQLFNWK8AovT1gFvda5jfw2oJeRMKEmw7aH6MGBJ3or",
];
/** Minimum tip for dual (SWQOS + Jito) submission — 0.001 SOL. */
const MIN_TIP_LAMPORTS_DUAL = 1000000n;
/** Minimum tip for SWQOS-only submission — 0.0005 SOL. */
const MIN_TIP_LAMPORTS_SWQOS = 500000n;
/** URL for the SFDP rejects list. */
const SFDP_REJECTS_URL = "https://helius-docs.s3.us-east-2.amazonaws.com/sfdp_rejects.json";

exports.MIN_TIP_LAMPORTS_DUAL = MIN_TIP_LAMPORTS_DUAL;
exports.MIN_TIP_LAMPORTS_SWQOS = MIN_TIP_LAMPORTS_SWQOS;
exports.SENDER_ENDPOINTS = SENDER_ENDPOINTS;
exports.SENDER_TIP_ACCOUNTS = SENDER_TIP_ACCOUNTS;
exports.SFDP_REJECTS_URL = SFDP_REJECTS_URL;
exports.senderFastUrl = senderFastUrl;
exports.senderPingUrl = senderPingUrl;
