// @helius-labs/helius-sdk/dist/cjs/staking/client.cjs
'use strict';

var createStakeTransaction = require('./createStakeTransaction.cjs');
var createUnstakeTransaction = require('./createUnstakeTransaction.cjs');
var createWithdrawTransaction = require('./createWithdrawTransaction.cjs');
var getHeliusStakeAccounts = require('./getHeliusStakeAccounts.cjs');
var getWithdrawableAmount = require('./getWithdrawableAmount.cjs');
var getStakeInstructions = require('./getStakeInstructions.cjs');
var getUnstakeInstruction = require('./getUnstakeInstruction.cjs');
var getWithdrawInstruction = require('./getWithdrawInstruction.cjs');

const makeStakeClientLazy = (rpc) => {
    const createStakeTransaction$1 = createStakeTransaction.makeCreateStakeTransaction({ rpc });
    const createUnstakeTransaction$1 = createUnstakeTransaction.makeCreateUnstakeTransaction({ rpc });
    const createWithdrawTransaction$1 = createWithdrawTransaction.makeCreateWithdrawTransaction({ rpc });
    const getHeliusStakeAccounts$1 = getHeliusStakeAccounts.makeGetHeliusStakeAccounts({ rpc });
    const getWithdrawableAmount$1 = getWithdrawableAmount.makeGetWithdrawableAmount({ rpc });
    const getStakeInstructions$1 = getStakeInstructions.makeGetStakeInstructions({ rpc });
    const getUnstakeInstruction$1 = getUnstakeInstruction.makeGetUnstakeInstruction();
    const getWithdrawInstruction$1 = getWithdrawInstruction.makeGetWithdrawInstruction();
    return {
        createStakeTransaction: createStakeTransaction$1,
        createUnstakeTransaction: createUnstakeTransaction$1,
        createWithdrawTransaction: createWithdrawTransaction$1,
        getHeliusStakeAccounts: getHeliusStakeAccounts$1,
        getWithdrawableAmount: getWithdrawableAmount$1,
        getStakeInstructions: getStakeInstructions$1,
        getUnstakeInstruction: getUnstakeInstruction$1,
        getWithdrawInstruction: getWithdrawInstruction$1,
    };
};

exports.makeStakeClientLazy = makeStakeClientLazy;
