// @helius-labs/helius-sdk/dist/cjs/staking/createWithdrawTransaction.cjs
'use strict';

var kit = require('@solana/kit');
var stake = require('@solana-program/stake');
var types = require('./types.cjs');

const makeCreateWithdrawTransaction = ({ rpc, }) => {
    return async (withdrawAuthority, stakeAccount, destination, lamports) => {
        const withdrawIx = stake.getWithdrawInstruction({
            stake: stakeAccount,
            recipient: destination,
            clockSysvar: types.SYSVAR_CLOCK,
            stakeHistory: types.SYSVAR_STAKE_HISTORY,
            withdrawAuthority,
            args: lamports,
        });
        const { value: blockHeight } = await rpc.getLatestBlockhash().send();
        const msg = kit.pipe(kit.createTransactionMessage({ version: 0 }), (m) => kit.setTransactionMessageFeePayer(withdrawAuthority.address, m), (m) => kit.setTransactionMessageLifetimeUsingBlockhash(blockHeight, m), (m) => kit.appendTransactionMessageInstructions([withdrawIx], m));
        const signedTx = await kit.signTransactionMessageWithSigners(msg);
        const serializedTx = kit.getBase64EncodedWireTransaction(signedTx);
        return { serializedTx };
    };
};

exports.makeCreateWithdrawTransaction = makeCreateWithdrawTransaction;
