// @helius-labs/helius-sdk/dist/cjs/staking/createStakeTransaction.cjs
'use strict';

var kit = require('@solana/kit');
var system = require('@solana-program/system');
var stake = require('@solana-program/stake');
var types = require('./types.cjs');

const makeCreateStakeTransaction = ({ rpc, }) => {
    return async (ownerSigner, amountSol) => {
        const rentExempt = await rpc
            .getMinimumBalanceForRentExemption(BigInt(types.STAKE_STATE_LEN))
            .send();
        const userLamports = BigInt(Math.round(amountSol * 1000000000));
        const lamports = rentExempt + userLamports;
        const stakeAccount = await kit.generateKeyPairSigner();
        const createIx = system.getCreateAccountInstruction({
            payer: ownerSigner,
            newAccount: stakeAccount,
            lamports,
            space: BigInt(types.STAKE_STATE_LEN),
            programAddress: types.STAKE_PROGRAM_ID,
        });
        const initializeIx = stake.getInitializeInstruction({
            stake: stakeAccount.address,
            arg0: { staker: ownerSigner.address, withdrawer: ownerSigner.address },
            arg1: { unixTimestamp: 0n, epoch: 0n, custodian: ownerSigner.address },
        });
        const delegateIx = stake.getDelegateStakeInstruction({
            stake: stakeAccount.address,
            vote: types.HELIUS_VALIDATOR_ID,
            clockSysvar: types.SYSVAR_CLOCK,
            stakeHistory: types.SYSVAR_STAKE_HISTORY,
            unused: types.UNUSED_STAKE_CONFIG_ACC,
            stakeAuthority: ownerSigner,
        });
        const { value: blockheight } = await rpc.getLatestBlockhash().send();
        const msg = kit.pipe(kit.createTransactionMessage({ version: 0 }), (m) => kit.setTransactionMessageFeePayer(ownerSigner.address, m), (m) => kit.setTransactionMessageLifetimeUsingBlockhash(blockheight, m), (m) => kit.appendTransactionMessageInstructions([createIx, initializeIx, delegateIx], m));
        const signedTx = await kit.signTransactionMessageWithSigners(msg);
        const serializedTx = kit.getBase64EncodedWireTransaction(signedTx);
        return {
            serializedTx,
            stakeAccountPubkey: stakeAccount.address,
        };
    };
};

exports.makeCreateStakeTransaction = makeCreateStakeTransaction;
