// @helius-labs/helius-sdk/dist/cjs/staking/getStakeInstructions.cjs
'use strict';

var kit = require('@solana/kit');
var system = require('@solana-program/system');
var stake = require('@solana-program/stake');
var types = require('./types.cjs');

const makeGetStakeInstructions = ({ rpc, }) => {
    return async (owner, amountSol) => {
        const rentExempt = await rpc
            .getMinimumBalanceForRentExemption(BigInt(types.STAKE_STATE_LEN))
            .send();
        const lamports = rentExempt + BigInt(Math.round(amountSol * 1000000000));
        const stakeAccount = await kit.generateKeyPairSigner();
        const createIx = system.getCreateAccountInstruction({
            payer: owner,
            newAccount: stakeAccount,
            lamports,
            space: BigInt(types.STAKE_STATE_LEN),
            programAddress: types.STAKE_PROGRAM_ID,
        });
        const initIx = stake.getInitializeInstruction({
            stake: stakeAccount.address,
            arg0: { staker: owner.address, withdrawer: owner.address },
            arg1: { unixTimestamp: 0n, epoch: 0n, custodian: owner.address },
        });
        const delegateIx = stake.getDelegateStakeInstruction({
            stake: stakeAccount.address,
            vote: types.HELIUS_VALIDATOR_ID,
            clockSysvar: types.SYSVAR_CLOCK,
            stakeHistory: types.SYSVAR_STAKE_HISTORY,
            unused: types.UNUSED_STAKE_CONFIG_ACC,
            stakeAuthority: owner,
        });
        return {
            instructions: [createIx, initIx, delegateIx],
            stakeAccount,
        };
    };
};

exports.makeGetStakeInstructions = makeGetStakeInstructions;
