// @helius-labs/helius-sdk/dist/esm/rpc/createHelius.eager.js
import { createSolanaRpcApi, DEFAULT_RPC_CONFIG, createDefaultRpcTransport, createRpc } from '@solana/kit';
import { wrapAutoSend } from './wrapAutoSend.js';
import { makeRpcCaller } from './caller.js';
import { getSDKHeaders } from '../http.js';
import { makeGetAsset } from './methods/getAsset.js';
import { makeGetAssetBatch } from './methods/getAssetBatch.js';
import { makeGetAssetProof } from './methods/getAssetProof.js';
import { makeGetAssetProofBatch } from './methods/getAssetProofBatch.js';
import { makeGetAssetsByAuthority } from './methods/getAssetsByAuthority.js';
import { makeGetAssetsByCreator } from './methods/getAssetsByCreator.js';
import { makeGetAssetsByGroup } from './methods/getAssetsByGroup.js';
import { makeGetPriorityFeeEstimate } from './methods/getPriorityFeeEstimate.js';
import { makeWebhookClientEager } from '../webhooks/client.eager.js';
import { makeWalletClientEager } from '../wallet/client.eager.js';
import { makeGetAssetsByOwner } from './methods/getAssetsByOwner.js';
import { makeGetNftEditions } from './methods/getNftEditions.js';
import { makeGetSignaturesForAsset } from './methods/getSignaturesForAsset.js';
import { makeGetTokenAccounts } from './methods/getTokenAccounts.js';
import { makeSearchAssets } from './methods/searchAssets.js';
import { makeEnhancedTxClientEager } from '../enhanced/client.eager.js';
import { makeTxHelpersEager } from '../transactions/client.eager.js';
import { makeGetProgramAccountsV2 } from './methods/getProgramAccountsV2.js';
import { makeGetAllProgramAccounts } from './methods/getAllProgramAccounts.js';
import { makeGetTokenAccountsByOwnerV2 } from './methods/getTokenAccountsByOwnerV2.js';
import { makeGetAllTokenAccountsByOwner } from './methods/getAllTokenAccountsByOwner.js';
import { makeGetTransactionsForAddress } from './methods/getTransactionsForAddress.js';

const createHeliusEager = ({ apiKey, network = "mainnet", rebateAddress, baseUrl, userAgent, }) => {
    // Use custom baseUrl if provided, otherwise construct from network
    const resolvedBaseUrl = baseUrl ?? `https://${network}.helius-rpc.com/`;
    // Build query parameters
    const queryParams = [];
    if (apiKey) {
        queryParams.push(`api-key=${apiKey}`);
    }
    if (rebateAddress) {
        queryParams.push(`rebate-address=${rebateAddress}`);
    }
    const queryString = queryParams.length > 0 ? `?${queryParams.join("&")}` : "";
    const url = `${resolvedBaseUrl}${queryString}`;
    const solanaApi = createSolanaRpcApi(DEFAULT_RPC_CONFIG);
    const transport = createDefaultRpcTransport({
        url,
        headers: getSDKHeaders(userAgent),
    });
    let baseRpc = createRpc({ api: solanaApi, transport });
    // Cast to any because I cba to go down this type rabbit hole
    const raw = wrapAutoSend(baseRpc);
    const call = makeRpcCaller(transport);
    return {
        raw,
        // DAS
        getAsset: makeGetAsset(call),
        getAssetBatch: makeGetAssetBatch(call),
        getAssetProof: makeGetAssetProof(call),
        getAssetProofBatch: makeGetAssetProofBatch(call),
        getAssetsByAuthority: makeGetAssetsByAuthority(call),
        getAssetsByCreator: makeGetAssetsByCreator(call),
        getAssetsByGroup: makeGetAssetsByGroup(call),
        getAssetsByOwner: makeGetAssetsByOwner(call),
        getNftEditions: makeGetNftEditions(call),
        getTokenAccounts: makeGetTokenAccounts(call),
        getSignaturesForAsset: makeGetSignaturesForAsset(call),
        searchAssets: makeSearchAssets(call),
        // Priority Fee API
        getPriorityFeeEstimate: makeGetPriorityFeeEstimate(call),
        // V2 methods
        getProgramAccountsV2: makeGetProgramAccountsV2(call),
        getAllProgramAccounts: makeGetAllProgramAccounts(call),
        getTokenAccountsByOwnerV2: makeGetTokenAccountsByOwnerV2(call),
        getAllTokenAccountsByOwner: makeGetAllTokenAccountsByOwner(call),
        getTransactionsForAddress: makeGetTransactionsForAddress(call),
        // Webhooks
        get webhooks() {
            if (!apiKey) {
                throw new Error("An API key is required to use webhooks/enhanced transactions. Provide apiKey in createHelius() options.");
            }
            return makeWebhookClientEager(apiKey, userAgent);
        },
        // Enhanced Transactions
        get enhanced() {
            if (!apiKey) {
                throw new Error("An API key is required to use webhooks/enhanced transactions. Provide apiKey in createHelius() options.");
            }
            return makeEnhancedTxClientEager(apiKey, network, userAgent);
        },
        // Transaction helpers
        tx: makeTxHelpersEager(baseRpc),
        // Wallet API
        get wallet() {
            if (!apiKey) {
                throw new Error("An API key is required to use the Wallet API. Provide apiKey in createHelius() options.");
            }
            return makeWalletClientEager(apiKey, userAgent);
        },
    };
};

export { createHeliusEager };
