// @helius-labs/helius-sdk/dist/cjs/auth/agenticSignup.cjs
'use strict';

var loadKeypair = require('./loadKeypair.cjs');
var getAddress = require('./getAddress.cjs');
var signAuthMessage = require('./signAuthMessage.cjs');
var walletSignup = require('./walletSignup.cjs');
var listProjects = require('./listProjects.cjs');
var getProject = require('./getProject.cjs');
var checkout = require('./checkout.cjs');
var constants = require('./constants.cjs');
var signupHelpers = require('./signupHelpers.cjs');
var basicSignup = require('./basicSignup.cjs');

async function agenticSignup(options) {
    const { secretKey, userAgent, email, firstName, lastName } = options;
    // Normalize plan: undefined/empty → "basic"
    const rawPlan = options.plan?.trim() || "";
    const plan = rawPlan === "" ? "basic" : rawPlan.toLowerCase();
    // Validate plan
    if (plan !== "basic" && !signupHelpers.isOpenPayPlan(plan)) {
        throw new Error(`Unknown plan: ${plan}. Available: basic, ${constants.OPENPAY_PLANS.join(", ")}`);
    }
    // Load keypair and derive address
    const keypair = loadKeypair.loadKeypair(secretKey);
    const walletAddress = await getAddress.getAddress(keypair);
    // Authenticate
    const { message, signature } = await signAuthMessage.signAuthMessage(secretKey);
    const auth = await walletSignup.walletSignup(message, signature, walletAddress, userAgent);
    const jwt = auth.token;
    // Check for existing projects
    const existingProjects = await listProjects.listProjects(jwt, userAgent);
    if (existingProjects.length > 0) {
        const project = existingProjects[0];
        const projectDetails = await getProject.getProject(jwt, project.id, userAgent);
        const apiKey = projectDetails.apiKeys?.[0]?.keyId || null;
        // Existing user + OpenPay plan → upgrade
        if (signupHelpers.isOpenPayPlan(plan)) {
            const upgradeResult = await checkout.executeUpgrade(secretKey, jwt, plan, options.period ?? "monthly", project.id, options.couponCode, userAgent);
            if (upgradeResult.status !== "completed") {
                throw new Error(`Checkout ${upgradeResult.status}${upgradeResult.error ? `: ${upgradeResult.error}` : ""}${upgradeResult.txSignature ? `. TX: ${upgradeResult.txSignature}` : ""}`);
            }
            return {
                status: "upgraded",
                jwt,
                walletAddress,
                projectId: project.id,
                apiKey,
                endpoints: apiKey ? signupHelpers.buildEndpoints(apiKey) : null,
                credits: null,
                txSignature: upgradeResult.txSignature ?? undefined,
            };
        }
        // Existing user + basic plan → return existing project
        return {
            status: "existing_project",
            jwt,
            walletAddress,
            projectId: project.id,
            apiKey,
            endpoints: apiKey ? signupHelpers.buildEndpoints(apiKey) : null,
            credits: projectDetails.creditsUsage?.remainingCredits ?? null,
        };
    }
    // ── New user paths ──
    if (signupHelpers.isOpenPayPlan(plan)) {
        // Validate required contact info for new subscriptions
        if (!email || !firstName || !lastName) {
            const missing = [
                !email && "email",
                !firstName && "firstName",
                !lastName && "lastName",
            ].filter(Boolean);
            throw new Error(`Paid plans require contact info for new accounts. Missing: ${missing.join(", ")}. ` +
                `Pass --email, --first-name, and --last-name.`);
        }
        // OpenPay checkout for developer/business/professional
        const checkoutResult = await checkout.executeCheckout(secretKey, jwt, {
            plan,
            period: options.period ?? "monthly",
            refId: auth.refId,
            email,
            firstName,
            lastName,
            couponCode: options.couponCode,
        }, userAgent);
        if (checkoutResult.status !== "completed") {
            throw new Error(`Checkout ${checkoutResult.status}${checkoutResult.error ? `: ${checkoutResult.error}` : ""}${checkoutResult.txSignature ? `. TX: ${checkoutResult.txSignature}` : ""}`);
        }
        if (!checkoutResult.projectId) {
            throw new Error("Checkout completed but no project was provisioned. " +
                `Payment intent: ${checkoutResult.paymentIntentId}`);
        }
        return {
            status: "success",
            jwt,
            walletAddress,
            projectId: checkoutResult.projectId,
            apiKey: checkoutResult.apiKey || null,
            endpoints: checkoutResult.apiKey
                ? signupHelpers.buildEndpoints(checkoutResult.apiKey)
                : null,
            credits: null,
            txSignature: checkoutResult.txSignature ?? undefined,
        };
    }
    // Basic plan ($1 USDC) → balance checks → pay → createProject
    return basicSignup.executeBasicSignup(secretKey, jwt, walletAddress, userAgent);
}

exports.agenticSignup = agenticSignup;
