// @helius-labs/helius-sdk/dist/cjs/auth/buildTokenTransfer.cjs
'use strict';

var kit = require('@solana/kit');
var token = require('@solana-program/token');
var constants = require('./constants.cjs');

async function buildAndSendTokenTransfer(params) {
    const { secretKey, recipientAddress, mintAddress, amount, additionalInstructions = [], } = params;
    const rpc = kit.createSolanaRpc(constants.RPC_URL);
    const rpcSubscriptions = kit.createSolanaRpcSubscriptions(constants.WS_URL);
    const sendAndConfirm = kit.sendAndConfirmTransactionFactory({
        rpc,
        rpcSubscriptions,
    });
    const signer = await kit.createKeyPairSignerFromBytes(secretKey);
    const [senderAta] = await token.findAssociatedTokenPda({
        owner: signer.address,
        mint: mintAddress,
        tokenProgram: token.TOKEN_PROGRAM_ADDRESS,
    });
    const [receiverAta] = await token.findAssociatedTokenPda({
        owner: kit.address(recipientAddress),
        mint: mintAddress,
        tokenProgram: token.TOKEN_PROGRAM_ADDRESS,
    });
    const { value: latestBlockhash } = await rpc.getLatestBlockhash().send();
    const transferIx = token.getTransferInstruction({
        source: senderAta,
        destination: receiverAta,
        authority: signer,
        amount,
    });
    const transactionMessage = kit.pipe(kit.createTransactionMessage({ version: 0 }), (tx) => kit.setTransactionMessageFeePayerSigner(signer, tx), (tx) => kit.setTransactionMessageLifetimeUsingBlockhash(latestBlockhash, tx), (tx) => kit.appendTransactionMessageInstructions([transferIx, ...additionalInstructions], tx));
    const signedTransaction = await kit.signTransactionMessageWithSigners(transactionMessage);
    const transactionWithBlockHeight = {
        ...signedTransaction,
        lifetimeConstraint: {
            lastValidBlockHeight: latestBlockhash.lastValidBlockHeight,
        },
    };
    await sendAndConfirm(transactionWithBlockHeight, { commitment: "confirmed" });
    return kit.getSignatureFromTransaction(signedTransaction);
}

exports.buildAndSendTokenTransfer = buildAndSendTokenTransfer;
