// @helius-labs/helius-sdk/dist/cjs/auth/client.cjs
'use strict';

var generateKeypair = require('./generateKeypair.cjs');
var loadKeypair = require('./loadKeypair.cjs');
var getAddress = require('./getAddress.cjs');
var signAuthMessage = require('./signAuthMessage.cjs');
var walletSignup = require('./walletSignup.cjs');
var listProjects = require('./listProjects.cjs');
var createProject = require('./createProject.cjs');
var getProject = require('./getProject.cjs');
var createApiKey = require('./createApiKey.cjs');
var checkBalances = require('./checkBalances.cjs');
var payUSDC = require('./payUSDC.cjs');
var payWithMemo = require('./payWithMemo.cjs');
var checkout = require('./checkout.cjs');
var agenticSignup = require('./agenticSignup.cjs');

function makeAuthClient(userAgent) {
    return {
        generateKeypair: generateKeypair.generateKeypair,
        loadKeypair: loadKeypair.loadKeypair,
        getAddress: getAddress.getAddress,
        signAuthMessage: signAuthMessage.signAuthMessage,
        walletSignup: (msg, sig, address) => walletSignup.walletSignup(msg, sig, address, userAgent),
        listProjects: (jwt) => listProjects.listProjects(jwt, userAgent),
        createProject: (jwt) => createProject.createProject(jwt, userAgent),
        getProject: (jwt, id) => getProject.getProject(jwt, id, userAgent),
        createApiKey: (jwt, projectId, wallet) => createApiKey.createApiKey(jwt, projectId, wallet, userAgent),
        checkSolBalance: checkBalances.checkSolBalance,
        checkUsdcBalance: checkBalances.checkUsdcBalance,
        payUSDC: payUSDC.payUSDC,
        initializeCheckout: (jwt, req) => checkout.initializeCheckout(jwt, req, userAgent),
        executeCheckout: (sk, jwt, req) => checkout.executeCheckout(sk, jwt, req, userAgent),
        payWithMemo: payWithMemo.payWithMemo,
        agenticSignup: (options) => agenticSignup.agenticSignup({ ...options, userAgent }),
        getCheckoutPreview: (jwt, plan, period, refId, coupon) => checkout.getCheckoutPreview(jwt, plan, period, refId, coupon, userAgent),
        getPaymentIntent: (jwt, id) => checkout.getPaymentIntent(jwt, id, userAgent),
        getPaymentStatus: (jwt, id) => checkout.getPaymentStatus(jwt, id, userAgent),
        payPaymentIntent: checkout.payPaymentIntent,
        executeUpgrade: (sk, jwt, plan, period, projectId, coupon) => checkout.executeUpgrade(sk, jwt, plan, period, projectId, coupon, userAgent),
        executeRenewal: (sk, jwt, id) => checkout.executeRenewal(sk, jwt, id, userAgent),
    };
}

exports.makeAuthClient = makeAuthClient;
