// @helius-labs/helius-sdk/dist/esm/staking/createWithdrawTransaction.js
import { pipe, createTransactionMessage, setTransactionMessageFeePayer, setTransactionMessageLifetimeUsingBlockhash, appendTransactionMessageInstructions, signTransactionMessageWithSigners, getBase64EncodedWireTransaction } from '@solana/kit';
import { getWithdrawInstruction } from '@solana-program/stake';
import { SYSVAR_STAKE_HISTORY, SYSVAR_CLOCK } from './types.js';

const makeCreateWithdrawTransaction = ({ rpc, }) => {
    return async (withdrawAuthority, stakeAccount, destination, lamports) => {
        const withdrawIx = getWithdrawInstruction({
            stake: stakeAccount,
            recipient: destination,
            clockSysvar: SYSVAR_CLOCK,
            stakeHistory: SYSVAR_STAKE_HISTORY,
            withdrawAuthority,
            args: lamports,
        });
        const { value: blockHeight } = await rpc.getLatestBlockhash().send();
        const msg = pipe(createTransactionMessage({ version: 0 }), (m) => setTransactionMessageFeePayer(withdrawAuthority.address, m), (m) => setTransactionMessageLifetimeUsingBlockhash(blockHeight, m), (m) => appendTransactionMessageInstructions([withdrawIx], m));
        const signedTx = await signTransactionMessageWithSigners(msg);
        const serializedTx = getBase64EncodedWireTransaction(signedTx);
        return { serializedTx };
    };
};

export { makeCreateWithdrawTransaction };
