// @helius-labs/helius-sdk/dist/esm/staking/getStakeInstructions.js
import { generateKeyPairSigner } from '@solana/kit';
import { getCreateAccountInstruction } from '@solana-program/system';
import { getInitializeInstruction, getDelegateStakeInstruction } from '@solana-program/stake';
import { STAKE_STATE_LEN, STAKE_PROGRAM_ID, UNUSED_STAKE_CONFIG_ACC, SYSVAR_STAKE_HISTORY, SYSVAR_CLOCK, HELIUS_VALIDATOR_ID } from './types.js';

const makeGetStakeInstructions = ({ rpc, }) => {
    return async (owner, amountSol) => {
        const rentExempt = await rpc
            .getMinimumBalanceForRentExemption(BigInt(STAKE_STATE_LEN))
            .send();
        const lamports = rentExempt + BigInt(Math.round(amountSol * 1000000000));
        const stakeAccount = await generateKeyPairSigner();
        const createIx = getCreateAccountInstruction({
            payer: owner,
            newAccount: stakeAccount,
            lamports,
            space: BigInt(STAKE_STATE_LEN),
            programAddress: STAKE_PROGRAM_ID,
        });
        const initIx = getInitializeInstruction({
            stake: stakeAccount.address,
            arg0: { staker: owner.address, withdrawer: owner.address },
            arg1: { unixTimestamp: 0n, epoch: 0n, custodian: owner.address },
        });
        const delegateIx = getDelegateStakeInstruction({
            stake: stakeAccount.address,
            vote: HELIUS_VALIDATOR_ID,
            clockSysvar: SYSVAR_CLOCK,
            stakeHistory: SYSVAR_STAKE_HISTORY,
            unused: UNUSED_STAKE_CONFIG_ACC,
            stakeAuthority: owner,
        });
        return {
            instructions: [createIx, initIx, delegateIx],
            stakeAccount,
        };
    };
};

export { makeGetStakeInstructions };
