// @helius-labs/helius-sdk/dist/esm/staking/types.js
/** Native Stake program address. */
const STAKE_PROGRAM_ID = "Stake11111111111111111111111111111111111111";
/** Sysvar Clock address. */
const SYSVAR_CLOCK = "SysvarC1ock11111111111111111111111111111111";
/** Sysvar Stake History address. */
const SYSVAR_STAKE_HISTORY = "SysvarStakeHistory1111111111111111111111111";
/** Unused stake config account (required by legacy instructions). */
const UNUSED_STAKE_CONFIG_ACC = "StakeConfig11111111111111111111111111111111";
/** Helius validator vote account address. */
const HELIUS_VALIDATOR_ID = "he1iusunGwqrNtafDtLdhsUQDFvo13z9sUa36PauBtk";
/** Max u64 value — active stake accounts have their deactivation epoch set to this. */
const U64_MAX = "18446744073709551615";
/** Size of a stake account in bytes. */
const STAKE_STATE_LEN = 200;
/** Lamports per SOL as a BigInt-compatible value. */
const LAMPORTS_PER_SOL_BIGINT = "1_000_000_000";

export { HELIUS_VALIDATOR_ID, LAMPORTS_PER_SOL_BIGINT, STAKE_PROGRAM_ID, STAKE_STATE_LEN, SYSVAR_CLOCK, SYSVAR_STAKE_HISTORY, U64_MAX, UNUSED_STAKE_CONFIG_ACC };
