// @helius-labs/helius-sdk/dist/cjs/types/enums.cjs
'use strict';

// Helper that keeps runtime values and infers the union type
const makeEnum = (o) => o;
/** Asset interface type — identifies the on-chain program standard. */
const Interface = makeEnum({
    V1_NFT: "V1_NFT",
    CUSTOM: "Custom",
    V1_PRINT: "V1_PRINT",
    LEGACY_NFT: "Legacy_NFT",
    V2_NFT: "V2_NFT",
    FUNGIBLE_ASSET: "FungibleAsset",
    IDENTITY: "Identity",
    EXECUTABLE: "Executable",
    PROGRAMMABLE_NFT: "ProgrammableNFT",
    FUNGIBLE_TOKEN: "FungibleToken",
    MPL_CORE_ASSET: "MplCoreAsset",
});
/** How ownership of an asset is modeled. */
const OwnershipModel = makeEnum({
    SINGLE: "single",
    TOKEN: "token",
});
/** Royalty distribution model for an asset. */
const RoyaltyModel = makeEnum({
    CREATORS: "creators",
    FANOUT: "fanout",
    SINGLE: "single",
});
/** Authority scope level for an asset. */
const Scope = makeEnum({
    FULL: "full",
    ROYALTY: "royalty",
    METADATA: "metadata",
    EXTENSION: "extension",
});
/** Method governing how an asset's uses are consumed. */
const UseMethods = makeEnum({
    BURN: "Burn",
    SINGLE: "Single",
    MULTIPLE: "Multiple",
});
/** Display context in which a file/asset is intended to be rendered. */
const Context = makeEnum({
    WALLET_DEFAULT: "wallet-default",
    WEB_DESKTOP: "web-desktop",
    WEB_MOBILE: "web-mobile",
    APP_MOBILE: "app-mobile",
    APP_DESKTOP: "app-desktop",
    APP: "app",
    VR: "vr",
});
/** Field by which DAS asset queries can be sorted. */
const AssetSortBy = makeEnum({
    ID: "id",
    CREATED: "created",
    UPDATED: "updated",
    RECENT_ACTION: "recent_action",
    NONE: "none",
});
/** Sort direction for DAS asset queries. */
const AssetSortDirection = makeEnum({ ASC: "asc", DESC: "desc" });
/** Token standard classification. */
const TokenStandard = makeEnum({
    PROGRAMMABLE_NON_FUNGIBLE: "ProgrammableNonFungible",
    NON_FUNGIBLE: "NonFungible",
    FUNGIBLE: "Fungible",
    FUNGIBLE_ASSET: "FungibleAsset",
    NON_FUNGIBLE_EDITION: "NonFungibleEdition",
    UNKNOWN_STANDARD: "UnknownStandard",
});
/** Priority fee level hint for the `getPriorityFeeEstimate` API. */
const PriorityLevel = makeEnum({
    MIN: "Min",
    LOW: "Low",
    MEDIUM: "Medium",
    HIGH: "High",
    VERY_HIGH: "VeryHigh",
    UNSAFE_MAX: "UnsafeMax",
    DEFAULT: "Default",
});
/** Transaction encoding format for UI / RPC requests. */
const UiTransactionEncoding = makeEnum({
    Binary: "binary",
    Base64: "base64",
    Base58: "base58",
    Json: "json",
    JsonParsed: "jsonParsed",
});

exports.AssetSortBy = AssetSortBy;
exports.AssetSortDirection = AssetSortDirection;
exports.Context = Context;
exports.Interface = Interface;
exports.OwnershipModel = OwnershipModel;
exports.PriorityLevel = PriorityLevel;
exports.RoyaltyModel = RoyaltyModel;
exports.Scope = Scope;
exports.TokenStandard = TokenStandard;
exports.UiTransactionEncoding = UiTransactionEncoding;
exports.UseMethods = UseMethods;
