// @helius-labs/helius-sdk/dist/esm/auth/buildTokenTransfer.js
import { createSolanaRpc, createSolanaRpcSubscriptions, sendAndConfirmTransactionFactory, createKeyPairSignerFromBytes, address, pipe, createTransactionMessage, setTransactionMessageFeePayerSigner, setTransactionMessageLifetimeUsingBlockhash, appendTransactionMessageInstructions, signTransactionMessageWithSigners, getSignatureFromTransaction } from '@solana/kit';
import { findAssociatedTokenPda, TOKEN_PROGRAM_ADDRESS, getTransferInstruction } from '@solana-program/token';
import { RPC_URL, WS_URL } from './constants.js';

async function buildAndSendTokenTransfer(params) {
    const { secretKey, recipientAddress, mintAddress, amount, additionalInstructions = [], } = params;
    const rpc = createSolanaRpc(RPC_URL);
    const rpcSubscriptions = createSolanaRpcSubscriptions(WS_URL);
    const sendAndConfirm = sendAndConfirmTransactionFactory({
        rpc,
        rpcSubscriptions,
    });
    const signer = await createKeyPairSignerFromBytes(secretKey);
    const [senderAta] = await findAssociatedTokenPda({
        owner: signer.address,
        mint: mintAddress,
        tokenProgram: TOKEN_PROGRAM_ADDRESS,
    });
    const [receiverAta] = await findAssociatedTokenPda({
        owner: address(recipientAddress),
        mint: mintAddress,
        tokenProgram: TOKEN_PROGRAM_ADDRESS,
    });
    const { value: latestBlockhash } = await rpc.getLatestBlockhash().send();
    const transferIx = getTransferInstruction({
        source: senderAta,
        destination: receiverAta,
        authority: signer,
        amount,
    });
    const transactionMessage = pipe(createTransactionMessage({ version: 0 }), (tx) => setTransactionMessageFeePayerSigner(signer, tx), (tx) => setTransactionMessageLifetimeUsingBlockhash(latestBlockhash, tx), (tx) => appendTransactionMessageInstructions([transferIx, ...additionalInstructions], tx));
    const signedTransaction = await signTransactionMessageWithSigners(transactionMessage);
    const transactionWithBlockHeight = {
        ...signedTransaction,
        lifetimeConstraint: {
            lastValidBlockHeight: latestBlockhash.lastValidBlockHeight,
        },
    };
    await sendAndConfirm(transactionWithBlockHeight, { commitment: "confirmed" });
    return getSignatureFromTransaction(signedTransaction);
}

export { buildAndSendTokenTransfer };
