// @helius-labs/helius-sdk/dist/esm/auth/checkout.js
import { authRequest, sleep } from './utils.js';
import { loadKeypair } from './loadKeypair.js';
import { getAddress } from './getAddress.js';
import { checkSolBalance, checkUsdcBalance } from './checkBalances.js';
import { payWithMemo } from './payWithMemo.js';
import { listProjects } from './listProjects.js';
import { getProject } from './getProject.js';
import { PLAN_TO_USAGE_PLAN, CHECKOUT_POLL_TIMEOUT_MS, CHECKOUT_POLL_INTERVAL_MS, MIN_SOL_FOR_TX, PROJECT_POLL_INTERVAL_MS, PROJECT_POLL_TIMEOUT_MS } from './constants.js';
import { fetchOpenPayPriceIds } from './devPortalConfigs.js';

async function resolvePriceId(jwt, plan, period, userAgent) {
    const usagePlan = PLAN_TO_USAGE_PLAN[plan.toLowerCase()];
    if (!usagePlan) {
        throw new Error(`Unknown plan: ${plan}. Available: ${Object.keys(PLAN_TO_USAGE_PLAN).join(", ")}`);
    }
    const priceIds = await fetchOpenPayPriceIds(jwt, userAgent);
    const periodKey = period === "monthly" ? "Monthly" : "Yearly";
    const priceId = priceIds[periodKey]?.[usagePlan];
    if (!priceId) {
        const available = Object.keys(priceIds[periodKey] ?? {});
        throw new Error(`No priceId found for plan "${plan}" (${period}). ` +
            (available.length === 0
                ? "The pricing configuration is empty — the backend may not be fully deployed yet."
                : `Expected key "${usagePlan}" but available keys are: [${available.join(", ")}]`));
    }
    return priceId;
}
async function initializeCheckout(jwt, request, userAgent) {
    return authRequest("/checkout/initialize", {
        method: "POST",
        headers: { Authorization: `Bearer ${jwt}` },
        body: JSON.stringify(request),
    }, userAgent);
}
async function getCheckoutPreview(jwt, plan, period, refId, couponCode, userAgent) {
    const priceId = await resolvePriceId(jwt, plan, period, userAgent);
    const params = new URLSearchParams({ priceId, refId });
    if (couponCode)
        params.set("couponCode", couponCode);
    return authRequest(`/checkout/preview?${params.toString()}`, {
        method: "GET",
        headers: { Authorization: `Bearer ${jwt}` },
    }, userAgent);
}
async function getPaymentIntent(jwt, paymentIntentId, userAgent) {
    return authRequest(`/checkout/${paymentIntentId}`, {
        method: "GET",
        headers: { Authorization: `Bearer ${jwt}` },
    }, userAgent);
}
async function getPaymentStatus(jwt, paymentIntentId, userAgent) {
    return authRequest(`/checkout/${paymentIntentId}/status`, {
        method: "GET",
        headers: { Authorization: `Bearer ${jwt}` },
    }, userAgent);
}
async function pollCheckoutCompletion(jwt, paymentIntentId, userAgent, options) {
    const timeoutMs = options?.timeoutMs ?? CHECKOUT_POLL_TIMEOUT_MS;
    const intervalMs = options?.intervalMs ?? CHECKOUT_POLL_INTERVAL_MS;
    const deadline = Date.now() + timeoutMs;
    while (Date.now() < deadline) {
        let status;
        try {
            status = await authRequest(`/checkout/${paymentIntentId}/status`, {
                method: "GET",
                headers: { Authorization: `Bearer ${jwt}` },
            }, userAgent);
        }
        catch (error) {
            // HTTP 410 Gone — intent expired
            if (error instanceof Error && error.message.includes("410")) {
                return {
                    status: "expired",
                    phase: "expired",
                    subscriptionActive: false,
                    readyToRedirect: false,
                    message: "Payment intent expired",
                };
            }
            throw error;
        }
        if (status.readyToRedirect) {
            return status;
        }
        if (status.phase === "failed" || status.phase === "expired") {
            return status;
        }
        await sleep(intervalMs);
    }
    return {
        status: "pending",
        phase: "confirming",
        subscriptionActive: false,
        readyToRedirect: false,
        message: "Polling timed out",
    };
}
async function payPaymentIntent(secretKey, intent) {
    // $0 intents are auto-completed by backend — no transaction needed
    if (intent.amount === 0) {
        return "";
    }
    const keypair = loadKeypair(secretKey);
    const walletAddress = await getAddress(keypair);
    const solBalance = await checkSolBalance(walletAddress);
    if (solBalance < MIN_SOL_FOR_TX) {
        throw new Error(`Insufficient SOL for transaction fees. Have: ${Number(solBalance) / 1000000000} SOL, need: ~0.001 SOL. Fund address: ${walletAddress}`);
    }
    // Convert cents → USDC 6-decimal raw: 4900 cents × 10,000 = 49,000,000 raw = 49.000000 USDC
    const amountRaw = BigInt(intent.amount) * 10000n;
    const usdcBalance = await checkUsdcBalance(walletAddress);
    if (usdcBalance < amountRaw) {
        throw new Error(`Insufficient USDC. Have: ${Number(usdcBalance) / 1000000} USDC, need: ${intent.amount / 100} USDC. Fund address: ${walletAddress}`);
    }
    // memo is intent.id
    return payWithMemo(secretKey, intent.destinationWallet, amountRaw, intent.id);
}
async function executeCheckout(secretKey, jwt, request, userAgent, options) {
    // 1. Resolve priceId from plan+period, then initialize checkout
    const priceId = await resolvePriceId(jwt, request.plan, request.period, userAgent);
    const intent = await initializeCheckout(jwt, {
        priceId,
        refId: request.refId,
        email: request.email,
        firstName: request.firstName,
        lastName: request.lastName,
        walletAddress: request.walletAddress,
        couponCode: request.couponCode,
    }, userAgent);
    // 2. Send payment (handles $0 case)
    let txSignature = null;
    try {
        txSignature = (await payPaymentIntent(secretKey, intent)) || null;
    }
    catch (error) {
        return {
            paymentIntentId: intent.id,
            txSignature: null,
            status: "failed",
            error: error instanceof Error ? error.message : String(error),
        };
    }
    // 3. Poll for payment confirmation
    const checkoutStatus = await pollCheckoutCompletion(jwt, intent.id, userAgent);
    if (checkoutStatus.phase === "failed") {
        return {
            paymentIntentId: intent.id,
            txSignature,
            status: "failed",
            error: checkoutStatus.message,
        };
    }
    if (checkoutStatus.phase === "expired") {
        return {
            paymentIntentId: intent.id,
            txSignature,
            status: "expired",
            error: checkoutStatus.message,
        };
    }
    if (!checkoutStatus.readyToRedirect) {
        return {
            paymentIntentId: intent.id,
            txSignature,
            status: "timeout",
        };
    }
    // 4. Optionally poll for project creation
    if (!options?.skipProjectPolling) {
        const projectDeadline = Date.now() + PROJECT_POLL_TIMEOUT_MS;
        let projectId;
        let apiKey;
        while (Date.now() < projectDeadline) {
            const projects = await listProjects(jwt, userAgent);
            if (projects.length > 0) {
                projectId = projects[0].id;
                const details = await getProject(jwt, projectId, userAgent);
                apiKey = details.apiKeys?.[0]?.keyId;
                break;
            }
            await sleep(PROJECT_POLL_INTERVAL_MS);
        }
        return {
            paymentIntentId: intent.id,
            txSignature,
            status: "completed",
            projectId,
            apiKey,
        };
    }
    return {
        paymentIntentId: intent.id,
        txSignature,
        status: "completed",
    };
}
async function executeUpgrade(secretKey, jwt, plan, period, projectId, couponCode, userAgent) {
    const priceId = await resolvePriceId(jwt, plan, period, userAgent);
    const intent = await initializeCheckout(jwt, { priceId, refId: projectId, couponCode }, userAgent);
    let txSignature = null;
    try {
        txSignature = (await payPaymentIntent(secretKey, intent)) || null;
    }
    catch (error) {
        return {
            paymentIntentId: intent.id,
            txSignature: null,
            status: "failed",
            error: error instanceof Error ? error.message : String(error),
        };
    }
    const status = await pollCheckoutCompletion(jwt, intent.id, userAgent);
    if (status.phase === "failed") {
        return {
            paymentIntentId: intent.id,
            txSignature,
            status: "failed",
            error: status.message,
        };
    }
    if (status.phase === "expired") {
        return {
            paymentIntentId: intent.id,
            txSignature,
            status: "expired",
            error: status.message,
        };
    }
    if (!status.readyToRedirect) {
        return { paymentIntentId: intent.id, txSignature, status: "timeout" };
    }
    return { paymentIntentId: intent.id, txSignature, status: "completed" };
}
async function executeRenewal(secretKey, jwt, paymentIntentId, userAgent) {
    const intent = await getPaymentIntent(jwt, paymentIntentId, userAgent);
    if (intent.status !== "pending") {
        throw new Error(`Payment intent is ${intent.status}, cannot pay. Only pending intents can be paid.`);
    }
    let txSignature = null;
    try {
        txSignature = (await payPaymentIntent(secretKey, intent)) || null;
    }
    catch (error) {
        return {
            paymentIntentId: intent.id,
            txSignature: null,
            status: "failed",
            error: error instanceof Error ? error.message : String(error),
        };
    }
    const status = await pollCheckoutCompletion(jwt, intent.id, userAgent);
    if (status.phase === "failed") {
        return {
            paymentIntentId: intent.id,
            txSignature,
            status: "failed",
            error: status.message,
        };
    }
    if (status.phase === "expired") {
        return {
            paymentIntentId: intent.id,
            txSignature,
            status: "expired",
            error: status.message,
        };
    }
    if (!status.readyToRedirect) {
        return { paymentIntentId: intent.id, txSignature, status: "timeout" };
    }
    return { paymentIntentId: intent.id, txSignature, status: "completed" };
}

export { executeCheckout, executeRenewal, executeUpgrade, getCheckoutPreview, getPaymentIntent, getPaymentStatus, initializeCheckout, payPaymentIntent, pollCheckoutCompletion, resolvePriceId };
