// @helius-labs/helius-sdk/dist/esm/auth/client.js
import { generateKeypair } from './generateKeypair.js';
import { loadKeypair } from './loadKeypair.js';
import { getAddress } from './getAddress.js';
import { signAuthMessage } from './signAuthMessage.js';
import { walletSignup } from './walletSignup.js';
import { listProjects } from './listProjects.js';
import { createProject } from './createProject.js';
import { getProject } from './getProject.js';
import { createApiKey } from './createApiKey.js';
import { checkUsdcBalance, checkSolBalance } from './checkBalances.js';
import { payUSDC } from './payUSDC.js';
import { payWithMemo } from './payWithMemo.js';
import { executeRenewal, executeUpgrade, payPaymentIntent, getPaymentStatus, getPaymentIntent, getCheckoutPreview, executeCheckout, initializeCheckout } from './checkout.js';
import { agenticSignup } from './agenticSignup.js';

function makeAuthClient(userAgent) {
    return {
        generateKeypair,
        loadKeypair,
        getAddress,
        signAuthMessage,
        walletSignup: (msg, sig, address) => walletSignup(msg, sig, address, userAgent),
        listProjects: (jwt) => listProjects(jwt, userAgent),
        createProject: (jwt) => createProject(jwt, userAgent),
        getProject: (jwt, id) => getProject(jwt, id, userAgent),
        createApiKey: (jwt, projectId, wallet) => createApiKey(jwt, projectId, wallet, userAgent),
        checkSolBalance,
        checkUsdcBalance,
        payUSDC,
        initializeCheckout: (jwt, req) => initializeCheckout(jwt, req, userAgent),
        executeCheckout: (sk, jwt, req) => executeCheckout(sk, jwt, req, userAgent),
        payWithMemo,
        agenticSignup: (options) => agenticSignup({ ...options, userAgent }),
        getCheckoutPreview: (jwt, plan, period, refId, coupon) => getCheckoutPreview(jwt, plan, period, refId, coupon, userAgent),
        getPaymentIntent: (jwt, id) => getPaymentIntent(jwt, id, userAgent),
        getPaymentStatus: (jwt, id) => getPaymentStatus(jwt, id, userAgent),
        payPaymentIntent,
        executeUpgrade: (sk, jwt, plan, period, projectId, coupon) => executeUpgrade(sk, jwt, plan, period, projectId, coupon, userAgent),
        executeRenewal: (sk, jwt, id) => executeRenewal(sk, jwt, id, userAgent),
    };
}

export { makeAuthClient };
