// @helius-labs/helius-sdk/dist/esm/auth/constants.js
const API_URL = "https://dev-api.helius.xyz/v0";
const TREASURY = "CEs84tEowsXpH8u4VBf8rJSVgSRypFMfXw9CpGRtQgb6";
const USDC_MINT_MAINNET = "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v";
const USDC_MINT_DEVNET = "4zMMC9srt5Ri5X14GAgXhaHii3GnPAEERYPJgZJDncDU";
const USDC_MINT = USDC_MINT_MAINNET;
/** Legacy: 1 USDC (6 decimals). Only used by payUSDC. */
const PAYMENT_AMOUNT = 1000000n;
/** Maps plan catalog keys to the keys returned by /dev-portal/configs openPay.priceIds */
const PLAN_TO_USAGE_PLAN = {
    developer: "developer_v4",
    business: "business_v4",
    professional: "professional_v4",
};
/** Minimum SOL needed for transaction fees (~0.001 SOL) */
const MIN_SOL_FOR_TX = 1000000n;
const MEMO_PROGRAM_ID = "MemoSq4gqABAXKb96qnH8TysNcWxMyWCqXgDLGmfcHr";
// ── OpenPay Plan Names (single source of truth) ──
const OPENPAY_PLANS = ["developer", "business", "professional"];
const CHECKOUT_POLL_INTERVAL_MS = 1000;
const CHECKOUT_POLL_TIMEOUT_MS = 60000;
const PROJECT_POLL_INTERVAL_MS = 2000;
const PROJECT_POLL_TIMEOUT_MS = 30000;
const RPC_URL = "https://api.mainnet-beta.solana.com";
const WS_URL = "wss://api.mainnet-beta.solana.com";

export { API_URL, CHECKOUT_POLL_INTERVAL_MS, CHECKOUT_POLL_TIMEOUT_MS, MEMO_PROGRAM_ID, MIN_SOL_FOR_TX, OPENPAY_PLANS, PAYMENT_AMOUNT, PLAN_TO_USAGE_PLAN, PROJECT_POLL_INTERVAL_MS, PROJECT_POLL_TIMEOUT_MS, RPC_URL, TREASURY, USDC_MINT, USDC_MINT_DEVNET, USDC_MINT_MAINNET, WS_URL };
